### PR TITLE
Add test case to return zero on a closed channel

### PR DESCRIPTION
**Pull request by kavyasrinet**
This test case covers the expected behavior of receiving in a closed channel as explained in the Axioms here : https://github.com/PaddlePaddle/Paddle/blob/2500f30ef210e8e55cccb5cc18ab525941e970f2/doc/design/csp.md

### CODE REVIEW

**Comment by abhinavarora** [comment]
@kavyasrinet I am not sure why we need this. @chengduoZH has already implemented a PR where the send and receive return a boolean. If channel got closed, we get a boolean false and if the send/receive was successful we get a true. I don't think this test is needed. We have already added tests for checking the boolean. This was done to keep the implementation similar to Go Channels and was also suggested by @wangkuiyi.  This is how Go implements this -> https://golang.org/ref/spec#Receive_operator. As long as we provide the boolean, the channel does not need to give any guarantees on the read value. It is the responsibility of the thread that is calling receive to see what boolean was returned. it should use the value in the pointer only if `true` was returned.

**Comment by kavyasrinet** [comment]
I agree with your comment above that closing a channel will return the boolean `false`, this test was written to explicitly check that a receive from a closed channel returns the zero value immediately, as is expected in a `Go` implementation and as was also mentioned in the axioms listed by @wangkuiyi here : https://github.com/PaddlePaddle/Paddle/blob/2500f30ef210e8e55cccb5cc18ab525941e970f2/doc/design/csp.md
This PR was intended to perform channel operations close to how `Go` does it.

File: paddle/framework/channel_test.cc [hunk 1]
```diff
@@ -60,6 +60,32 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
   delete ch;
 }
 
+// This test tests that CloseChannel returns a value of zero
+// immediately to all receivers that are trying to receive from the channel.
+TEST(Channel, ReceiverGetsZeroOnClosedChannel) {
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: If the purpose of this PR is to add a unit test to ensure that 

1. receiving from a closed buffered channel returns residual values and then zeros values, and
1. receiving from a closed unbuffered channel doesn't block and returns zero.

then this PR looks too simple to ensure the above two invariants.

Comment (kavyasrinet) [reply paddle/framework/channel_test.cc]: I think one of the axioms was  that : `A receive from a closed channel returns the residual values and then zeros.`
This PR tries to validate that #1 you explained above holds since we aren't explicitly checking for the zero value anywhere else.

File: paddle/framework/channel_test.cc [hunk 2]
```diff
@@ -60,6 +60,32 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+  const size_t buffer_size = 10;
+  auto ch = MakeChannel<size_t>(buffer_size);
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: This invariant should be verified with buffered and unbuffered channels.

File: paddle/framework/channel_test.cc [hunk 3]
```diff
@@ -60,6 +60,32 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+
+  for (size_t i = 1; i <= buffer_size; ++i) {
+    EXPECT_EQ(ch->Send(&i), true);  // sending should not block
+  }
+
+  CloseChannel(ch);
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: Maybe receive some values before CloseChannel and checks that receivings after CloseChannel return the residual values.

File: paddle/framework/channel_test.cc (hunk 2 above)
Comment (kavyasrinet) [reply paddle/framework/channel_test.cc]: On a side note, correct me if I am wrong @wangkuiyi but for an unbuffered channel, since the buffer size is zero, there would be no residual values. So trying to read from a closed channel should just return a zero value, right ?

**Comment by wangkuiyi** [comment]
https://github.com/PaddlePaddle/Paddle/pull/8175 This is for your reference.

**Comment by kavyasrinet** [comment]
Thank you!
I just pushed similar changes in this PR too. I will wait for the CI to finish running on #8175 and we can merge it.

File: paddle/framework/channel_test.cc [hunk 4]
```diff
@@ -62,27 +62,32 @@ TEST(Channel, SufficientBufferSizeDoesntBlock) {
...
+    EXPECT_EQ(out, i);
+  }
   CloseChannel(ch);
 
   // Now try receiving for more number of times than buffer size
   // after channel is closed
-  int out;
-  for (size_t i = 1; i <= 12; ++i) {
+
+  for (size_t i = 6; i <= 12; ++i) {
```
Comment (wangkuiyi) [inline paddle/framework/channel_test.cc]: Both 6 and 12 are magical numbers.

**Comment by kavyasrinet** [comment]
Closing this, since #8175 fixes the same issue.
